File: BARCODE-API/build_starter_db.py
import sys
import io
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8', errors='replace')

import urllib.request
import urllib.parse
import json
import sqlite3
import time
import os

OUTPUT_JSON = os.path.join(os.path.dirname(os.path.abspath(__file__)), "ph_grocery_starter.json")
OUTPUT_SQLITE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "starter_db.sqlite")
PAGE_SIZE = 100
MAX_PAGES = 30

def fetch_ph_products():
    print("[START] Starting Open Food Facts Philippine Dataset Scraper...\n")
    
    all_products = []
    barcode_set = set()

    for page in range(1, MAX_PAGES + 1):
        print(f"fetching Page {page} of {MAX_PAGES}...")

        params = urllib.parse.urlencode({
            "action": "process",
            "tagtype_0": "countries",
            "tag_contains_0": "contains",
            "tag_0": "philippines",
            "page_size": PAGE_SIZE,
            "page": page,
            "json": "true",
            "fields": "code,product_name,brands"
        })
        url = f"https://world.openfoodfacts.org/cgi/search.pl?{params}"

        req = urllib.request.Request(url, headers={
            "User-Agent": "PH-POS-StarterPackBuilder/1.0 (offline-pos-dev)"
        })

        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                if resp.status != 200:
                    print(f"[WARN] Page {page} returned status {resp.status}. Skipping...")
                    continue

                data = json.loads(resp.read().decode("utf-8"))
                products = data.get("products", [])

                if len(products) == 0:
                    print("[INFO] No more products found. Stopping fetch loop.")
                    break

                for item in products:
                    barcode = (item.get("code") or "").strip()
                    product_name = (item.get("product_name") or "").strip()
                    brand = (item.get("brands") or "").strip()

                    if barcode and product_name:
                        if barcode not in barcode_set:
                            barcode_set.add(barcode)
                            all_products.append({
                                "barcode": barcode,
                                "product_name": product_name,
                                "brand": brand
                            })

                time.sleep(0.3)

        except Exception as err:
            print(f"[ERROR] fetching page {page}: {err}")

    print(f"\n[DONE] Total unique Philippine products scraped: {len(all_products)}")

    with open(OUTPUT_JSON, "w", encoding="utf-8") as f:
        json.dump(all_products, f, indent=2, ensure_ascii=False)
    print(f"[SAVE] Saved JSON to: {OUTPUT_JSON}")

    db = sqlite3.connect(OUTPUT_SQLITE)
    cur = db.cursor()
    cur.executescript("""
        CREATE TABLE IF NOT EXISTS products (
            barcode TEXT PRIMARY KEY,
            product_name TEXT NOT NULL,
            brand TEXT
        );
        CREATE INDEX IF NOT EXISTS idx_barcode ON products(barcode);
    """)
    cur.executemany(
        "INSERT OR REPLACE INTO products (barcode, product_name, brand) VALUES (?, ?, ?)",
        [(p["barcode"], p["product_name"], p["brand"]) for p in all_products]
    )
    db.commit()
    db.close()
    print(f"[SAVE] Saved SQLite DB to: {OUTPUT_SQLITE}")
    print("\n[SUCCESS] Done! Ready to embed in your Offline POS system.")

if __name__ == "__main__":
    fetch_ph_products()

File: BARCODE-API/build_starter_db.js
const fs = require('fs');
const path = require('path');
const Database = require('better-sqlite3');

// Configuration
const OUTPUT_JSON = path.join(__dirname, 'ph_grocery_starter.json');
const OUTPUT_SQLITE = path.join(__dirname, 'starter_db.sqlite');
const PAGE_SIZE = 100; // Maximum allowed per page by OFF
const MAX_PAGES = 30;  // 30 pages * 100 = ~3,000 items

async function fetchPHProducts() {
  console.log('🚀 Starting Open Food Facts Philippine Dataset Scraper...\n');
  
  let allProducts = [];
  let barcodeSet = new Set(); // To prevent duplicate barcodes

  for (let page = 1; page <= MAX_PAGES; page++) {
    console.log(`fetching Page ${page} of ${MAX_PAGES}...`);
    
    // Search products with country=philippines or barcode prefix starting with 480
    const url = `https://world.openfoodfacts.org/cgi/search.pl?action=process&tagtype_0=countries&tag_contains_0=contains&tag_0=philippines&page_size=${PAGE_SIZE}&page=${page}&json=true&fields=code,product_name,brands`;

    try {
      const response = await fetch(url, {
        headers: { 'User-Agent': 'PH-POS-StarterPackBuilder/1.0 (offline-pos-dev)' }
      });

      if (!response.ok) {
        console.warn(`⚠️ Warning: Page ${page} returned status ${response.status}. Skipping...`);
        continue;
      }

      const data = await response.json();
      const products = data.products || [];

      if (products.length === 0) {
        console.log('ℹ️ No more products found. Stopping fetch loop.');
        break;
      }

      for (const item of products) {
        const barcode = item.code?.trim();
        const productName = item.product_name?.trim();
        const brand = item.brands?.trim() || '';

        // Filter: Must have a barcode and product name
        if (barcode && productName) {
          // Extra check: Accept 480 prefix or any valid code from PH search
          if (!barcodeSet.has(barcode)) {
            barcodeSet.add(barcode);
            allProducts.push({
              barcode: barcode,
              product_name: productName,
              brand: brand
            });
          }
        }
      }

      // Small delay to be respectful to OFF servers
      await new Promise(resolve => setTimeout(resolve, 300));

    } catch (err) {
      console.error(`❌ Error fetching page ${page}:`, err.message);
    }
  }

  console.log(`\n✅ Total unique Philippine products scraped: ${allProducts.length}`);

  // 1. Save to JSON File
  fs.writeFileSync(OUTPUT_JSON, JSON.stringify(allProducts, null, 2), 'utf-8');
  console.log(`💾 Saved JSON to: ${OUTPUT_JSON}`);

  // 2. Save to SQLite Database
  const db = new Database(OUTPUT_SQLITE);
  
  // Create Table
  db.exec(`
    CREATE TABLE IF NOT EXISTS products (
      barcode TEXT PRIMARY KEY,
      product_name TEXT NOT NULL,
      brand TEXT
    );
    CREATE INDEX IF NOT EXISTS idx_barcode ON products(barcode);
  `);

  // Insert items using Transaction (Fast batch insertion)
  const insertStmt = db.prepare(`
    INSERT OR REPLACE INTO products (barcode, product_name, brand) 
    VALUES (@barcode, @product_name, @brand)
  `);

  const insertMany = db.transaction((items) => {
    for (const item of items) insertStmt.run(item);
  });

  insertMany(allProducts);
  db.close();

  console.log(`💾 Saved SQLite DB to: ${OUTPUT_SQLITE}`);
  console.log('\n🎉 Done! Ready to embed in your Offline POS system.');
}

fetchPHProducts();
